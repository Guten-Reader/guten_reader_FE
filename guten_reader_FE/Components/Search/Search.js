import React, { Component } from 'react';
import { StyleSheet, View, Button, ScrollView, Text, TextInput } from 'react-native';
import ListSearch from '../ListSearch/ListSearch'

class Search extends Component {
  constructor() {
    super();
    this.state = {
      searchQuery: '',
      foundBooks: [],
      searchResult: false,
    }
  }

  updateState = (enteredText) => {
    this.setState({ searchQuery: enteredText})
  }

  searchBtn = async () => {
    if (this.state.searchQuery) {
      const response = await fetch(`http://gutendex.com/books?search=${this.state.searchQuery}`)
      try {
        const data = await response.json();
        this.filterContent(data.results);
        this.setState({ searchResult: true })
      }
      catch {
        this.setState({ foundBooks: [] })
      }
    } else {
      this.setState({ searchResult: false })
    }
  }

  filterContent = (data) => {
    if (data.length === 0) {
      console.log("Data length is 0");
      this.setState({ searchResult: false })
      this.setState({ foundBooks: [] })
      return;
    }
    let filterByMediaTypes = data.filter(book => book.media_type === 'Text')
    let cleanedBooks = filterByMediaTypes.map(book => {
      return {
        id: book.id,
        title: book.title,
        author: book.authors[0].name
      }
    });
    this.setState({ foundBooks: cleanedBooks })
  }

  render() {
    let renderSearchResults = [];
    let searchResult = 'No results from search, try again'
    if (this.state.foundBooks.length > 0) {
      renderSearchResults = this.state.foundBooks.map(book => {
        return <ListSearch book={book}/>
      })
    }
    return (
      <View style={styles.container}>
        <Text>Search by Title or Author:</Text>
        <TextInput
          placeholder="Title or Author"
          style={styles.input}
          onChangeText={this.updateState}
          value={this.state.searchQuery}
        />
        <Button title="SEARCH" onPress={this.searchBtn}/>
        {this.state.searchResult ? null : <Text>{searchResult}</Text>}
        <ScrollView>
          {renderSearchResults}
        </ScrollView>
        <Button style={styles.button} onPress={() => this.props.navigation.navigate('Library')} title="BACK"></Button>
      </View>
    )
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
  input: {
    borderColor: 'black',
  }
})

export default Search;
