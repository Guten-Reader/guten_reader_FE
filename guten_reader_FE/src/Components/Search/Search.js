import React, { Component } from 'react';
import { StyleSheet, View, Button, ScrollView, Text, TextInput, StatusBar } from 'react-native';
import ListSearch from '../ListSearch/ListSearch';
import Icon from 'react-native-vector-icons/SimpleLineIcons';

class Search extends Component {
  static navigationOptions = {
    title: 'Search',
    headerStyle: {
       backgroundColor: '#000000',
     },
    headerTintColor: '#53E69B',
    headerTitleStyle: {
      color: '#FFFFFF',
    }
  }

  constructor() {
    super();
    this.state = {
      searchQuery: '',
      foundBooks: [],
      searchResult: true,
    }
  }

  updateState = (enteredText) => {
    this.setState({ searchQuery: enteredText})
  }

  searchBtn = async () => {
    if (this.state.searchQuery) {
      const response = await fetch(`http://gutendex.com/books?search=${this.state.searchQuery}`)
      const data = await response.json();
      this.filterContent(data.results);
    }
  }

  filterContent = (data) => {
    if (data.length === 0) {
      this.setState({ searchResult: false })
      this.setState({ foundBooks: [] })
      return;
    }
    let filterByMediaType = data.filter(book => book.media_type === 'Text')
    let cleanedBooks = this.mapCleanBooks(filterByMediaType)
    this.setState({ searchResult: true })
    this.setState({ foundBooks: cleanedBooks })
  }

  mapCleanBooks = (books) => {
    return books.map(book => {
      let author = "unknown author"
      if (book.authors[0] !== undefined) {
        author = book.authors[0].name
      }
      return {
        id: book.id,
        title: book.title,
        author: author
      }
    });
  }

  render() {
    let renderSearchResults = [];
    let searchError = 'No results from search, try again';
    if (this.state.foundBooks.length > 0) {
      renderSearchResults = this.state.foundBooks.map((book, index) => {
        return <ListSearch book={book} key={index}/>
      })
    }

    return (
      <View style={styles.container}>
      <StatusBar barStyle="light-content" />
        <Text style={styles.title}>Search by Title or Author:</Text>
        <View style={styles.search}>
          <TextInput
            placeholder="Title or Author"
            style={styles.input}
            onChangeText={this.updateState}
            value={this.state.searchQuery}
          />
          <Icon style={styles.magnifier} name="magnifier" color="#53E69B" onPress={this.searchBtn} />
        </View>
        <View style={{ borderBottomColor: 'grey', borderBottomWidth: 1, }} />
        {this.state.searchResult ? null : <Text style={styles.title}>{searchError}</Text>}
        <ScrollView style={styles.scrollview}>
          {renderSearchResults}
        </ScrollView>
      </View>
    )
  }
}

const styles = StyleSheet.create({
  container: {
    backgroundColor: 'black',
    flex: 1,
    flexDirection: 'column'
  },
  title: {
   color: 'white',
   fontSize: 20,
   fontWeight: 'bold',
   marginTop: 20,
   marginRight: 20,
   marginLeft: 20,
   marginBottom: 10,
 },
 scrollview: {
   color: 'white',
   margin: 20
 },
 input: {
   color: 'black',
   backgroundColor: 'white',
   borderColor: 'white',
   borderWidth: 1,
   marginTop: 10,
   marginBottom: 10,
   padding: 15,
   marginLeft: 20,
   width: '77%'
 },
 magnifier: {
   fontSize: 25,
   borderWidth: 1,
   borderColor: '#d6d7da',
   padding: 9.4,
 },
 search: {
   alignItems: 'center',
   flexDirection: 'row',
   marginBottom: 20
 }
})

export default Search;
