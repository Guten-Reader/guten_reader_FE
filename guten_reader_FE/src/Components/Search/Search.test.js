import React from 'react';
// import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';

import Search from './Search';

describe('Search renders correctly with Jest', () => {
  it('renders correctly', () => {
    const wrapper = shallow(<Search />);
    expect(wrapper).toMatchSnapshot();
  })
})