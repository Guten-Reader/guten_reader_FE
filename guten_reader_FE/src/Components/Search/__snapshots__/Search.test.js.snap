// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search renders correctly with Jest renders correctly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "black",
      "flex": 1,
      "flexDirection": "column",
    }
  }
>
  <Text
    style={
      Object {
        "color": "white",
        "fontSize": 15,
        "fontWeight": "bold",
        "marginBottom": 10,
        "marginLeft": 20,
        "marginRight": 20,
        "marginTop": 40,
      }
    }
  >
    Search by Title or Author:
  </Text>
  <TextInput
    allowFontScaling={true}
    onChangeText={[Function]}
    placeholder="Title or Author"
    rejectResponderTermination={true}
    style={
      Object {
        "backgroundColor": "white",
        "borderColor": "white",
        "borderWidth": 1,
        "color": "black",
        "marginBottom": 10,
        "marginLeft": 20,
        "marginRight": 20,
        "marginTop": 10,
        "padding": 15,
      }
    }
    underlineColorAndroid="transparent"
    value=""
  />
  <Button
    onPress={[Function]}
    title="SEARCH"
  />
  <Text
    style={
      Object {
        "color": "white",
        "fontSize": 15,
        "fontWeight": "bold",
        "marginBottom": 10,
        "marginLeft": 20,
        "marginRight": 20,
        "marginTop": 40,
      }
    }
  >
    No results from search, try again
  </Text>
  <ScrollViewMock
    style={
      Object {
        "color": "white",
        "margin": 20,
      }
    }
  />
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#53E69B",
        "bottom": 0,
        "height": 65,
        "justifyContent": "center",
        "left": 0,
        "position": "absolute",
        "right": 0,
        "width": "100%",
      }
    }
  >
    <Button
      onPress={[Function]}
      title="BACK"
    />
  </View>
</View>
`;
